Convert Fullstack Tutorial to use TypeScript

stemmlerjs opened the pull request:
I'm working on converting the full-stack tutorial to TypeScript and here's where we are so far:

1. ✅ Fork the tutorial and rewrite the client portion in TS
2. ✅ Pair with @trevorblades to convert the first code block in the client tutorial over to TS

> ~~If this **goes well**, continue. If it **doesn’t go well**, sync w/ @peggyrayzis  for the next steps.~~ It went exceptionally well. Onwards.

3. Work with @trevorblades to make language settings global for the page (a user should be able to select TypeScript as the language of choice to read the tutorial and then they should see all the code samples in that language).

4. ✅Fork the docs and make changes to the code samples
5. Work with @StephenBarlow to perform minor copy changes if need be.

apollo-cla commented:
@stemmlerjs: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

stemmlerjs commented:
Just dropping this here as well 😺 

This is another issue that we might want to have resolved before we roll this out.
https://github.com/apollographql/gatsby-theme-apollo/issues/77

stemmlerjs commented:
Here's the PR from the Docs repo that uses the code from _this PR_ for code samples.

peggyrayzis commented on final/client/src/containers/cart-item.tsx:
```diff
@@ -14,7 +15,9 @@ export const GET_LAUNCH = gql`
-export default function CartItem({ launchId }) {
+interface CartItemProps extends LaunchDetailTypes.LaunchDetailsVariables {}
+
+const CartItem: React.FC<CartItemProps> = ({ launchId }) => {
```
What's your thought process behind extending the `LaunchDetailTypes.LaunchDetailsVariables` interface here? If there were additional props we had to declare, I could see why you would. Since there aren't additional props, we should be able to type `CartItem` as `React.FC<LaunchDetailTypes.LaunchDetailsVariables>`.

stemmlerjs replied:
@peggyrayzis My reasoning was to keep the naming structure of props for a component consistent (like what was done for `LaunchTile`, `LaunchDetails`, etc).

My personal preference is that I like to ship a component _with_ the type or interface that describes its inputs, _composed_ by types from the models (or in this case, the query variables).

Your suggestion is functionally equivalent to what I have here so we could do either one.

peggyrayzis replied:
Awesome! On second thought, I like the separation of concerns. Let's go with it for all other components.

peggyrayzis commented on final/client/src/containers/cart-item.tsx:
```diff
@@ -14,7 +15,9 @@ export const GET_LAUNCH = gql`
+interface CartItemProps extends LaunchDetailTypes.LaunchDetailsVariables {}
+
+const CartItem: React.FC<CartItemProps> = ({ launchId }) => {
   const { data, loading, error } = useQuery(
```
type `useQuery` hook

stemmlerjs replied:
Thanks :)

peggyrayzis commented on final/client/src/pages/cart.tsx:
```diff
@@ -4,30 +4,43 @@ import gql from 'graphql-tag';
+    data, 
+    loading, 
+    error 
+  }: QueryResult<GetCartItems, any> = useQuery<GetCartItems, any>(GET_CART_ITEMS);
```
I think you can type this hook as `const { data, loading, error } = useQuery<GetCartItems>(GET_CART_ITEMS);`. No need to declare any or specify `QueryResult`.

stemmlerjs replied:
Fixed.

peggyrayzis commented on final/client/src/pages/launch.tsx:
```diff
@@ -0,0 +1,58 @@
+  }: QueryResult<
+    LaunchDetailsTypes.LaunchDetails, 
+    LaunchDetailsTypes.LaunchDetailsVariables
+  > = useQuery(
```
No need to specify `QueryResult`. You can type this as:

```ts
const { data, loading, error } = useQuery<
    LaunchDetailsTypes.LaunchDetails,
    LaunchDetailsTypes.LaunchDetailsVariables
>(GET_LAUNCH_DETAILS, { variables: { launchId } });
```

stemmlerjs replied:
Fixed!

peggyrayzis commented on final/client/src/pages/launches.tsx:
```diff
@@ -33,17 +36,28 @@ export const GET_LAUNCHES = gql`
+    loading, 
+    error, 
+    fetchMore 
+  }: QueryResult<
```
See the feedback above on how I typed other `useQuery` calls.

peggyrayzis commented on final/client/src/pages/profile.tsx:
```diff
@@ -18,19 +20,22 @@ export const GET_MY_TRIPS = gql`
+interface ProfileProps extends RouteComponentProps {}
+
+const Profile: React.FC<ProfileProps> = () => {
+  const { data, loading, error } = useQuery<GetMyTripsTypes.GetMyTrips, any>(
```
no need to specify `any` for the variables when typing `useQuery`

peggyrayzis commented on final/client/src/resolvers.tsx:
```diff
@@ -0,0 +1,64 @@
+export const resolvers: AppResolvers = {
+  Launch: {
+    isInCart: (launch: LaunchTileTypes.LaunchTile, _, { cache }): boolean => {
+      const queryResult = cache.readQuery<GetCartItemTypes.GetCartItems, any>({ query: GET_CART_ITEMS });
```
remove `any`

stemmlerjs replied:
Got it!

peggyrayzis commented on final/client/src/resolvers.tsx:
```diff
@@ -0,0 +1,64 @@
+  },
+  Mutation: {
+    addOrRemoveFromCart: (_, { id }: { id: string }, { cache }): string[] => {
+      const queryResult = cache.readQuery<GetCartItemTypes.GetCartItems, any>({ query: GET_CART_ITEMS });
```
remove `any`

stemmlerjs replied:
Got this one as well.

peggyrayzis commented on start/client/package.json:
```diff
@@ -11,17 +11,20 @@
     "emotion": "^9.2.12",
     "graphql": "^14.4.2",
     "graphql-tag": "^2.10.1",
+    "lodash": "^4.17.15",
```
remove `lodash` -- I don't see where we're using it

peggyrayzis commented on start/client/package.json:
```diff
@@ -33,9 +36,16 @@
     "@testing-library/react": "^8.0.7",
-    "apollo": "^2.16.3"
+    "@types/jest": "^24.0.23",
+    "@types/lodash": "^4.14.149",
```
remove `lodash`

peggyrayzis commented on start/client/tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
I don't think we need to target `es5` since this tutorial is targeted at developers with modern browsers. Is this the generic `tsconfig.json` generated by CRA?

stemmlerjs replied:
@peggyrayzis Yeah, CRA automatically adds the `tsconfig.json` with these settings for us.
https://create-react-app.dev/docs/adding-typescript/

peggyrayzis commented:
@hwillson -- after @stemmlerjs is done addressing my feedback, could you please give this PR a review to make sure everything is in line with AC/TypeScript best practices?

peggyrayzis commented:
This is a style nit -- I noticed you changed the way we define React components from function declarations to function expressions with `const`. The React docs have moved toward using function declarations, and I'd like us to also be aligned with their recommendations.

I'm guessing your change has something to do with TypeScript. Is it worth changing them back to function declarations? @trevorblades and @hwillson -- what do you think?

stemmlerjs commented on final/client/src/components/launch-detail.tsx:
```diff
@@ -3,15 +3,18 @@ import styled from 'react-emotion';
+import { LaunchDetails_launch } from '../pages/__generated__/LaunchDetails';
 
-const LaunchDetail = ({ id, site, rocket }) => (
+type LaunchDetailProps = Partial<LaunchDetails_launch>
```
@peggyrayzis Good question! The reason we're using a `Partial` is:
- we want to add strict typing to the props of the `LaunchDetail` component
- but we don't need _all_ the props from a `LaunchDetails_launch` query. 

The way to tell TypeScript that we're going to use _some_ of the props from a type or interface is to use the `Partial<T>` type.

Without this, we'd get errors the way that we're destructuring `{ id, site, rocket }` from `LaunchDetail`.

stemmlerjs commented on final/client/package.json:
```diff
@@ -37,6 +42,8 @@
     "@testing-library/jest-dom": "^4.0.0",
     "@testing-library/react": "^8.0.7",
+    "@types/jest": "^24.0.23",
+    "@types/lodash": "^4.14.149",
```
Good eye. Thanks!

hwillson commented:
Thanks @peggyrayzis - @stemmlerjs, ping me when you're ready for another review. As for function declarations vs. function expressions, I've been leaning more towards function declarations as well, to align with React and other projects.

stemmlerjs commented:
Alrighty. Here's what's been updated.
- Updated the code based on @peggyrayzis's feedback. There are a couple more unresolved code blocks to come to a resolution on.
- Changed the startup script so that it would run the apollo codegen in watch mode as well. So `npm run start` will do `concurrently \"react-scripts start\" \"npm-watch\"`.

Here are some concerns:

A) When running the apollo codegen on the final version of the code, the `addOrRemoveFromCart` mutation stops the codegen from running. I think that it thinks that the mutation is a query and it's trying to create the types for it. It'll fail because there's nothing in the selection set.

```bash
✔ Loading Apollo Project
  ✖ Generating query files with 'typescript' target
    → Validation of GraphQL query document failed
.../client/src/containers/action-button.tsx: Field "addOrRemoveFromCart" of type "[Launch]" must have a selection of subfields. Did you mean "addOrRemoveFromCart { ... }"?
{ ToolError: Validation of GraphQL query document failed
    at Object.validateQueryDocument (/Users/khalilstemmler/Documents/Apollo/Code/fullstack-tutorial/final/client/node_modules/apollo-language-server/lib/errors/validation.js:29:19)
```

That's probably a very common issue for people. It only works when we comment out the `addOrRemoveFromCart` mutation. 

Several options for how to handle this:
- Keep that mutation commented out until we need it, though the console will be throwing non-exiting errors.
- Fix the codegen issue (not sure how long that will take, maybe someone from tooling can help). I can also take a look at this.
- Don't run the codegen on save

B) Do we want to go ahead with changing all the function expressions to function declarations? Let me know if the following format is preferable @peggyrayzis.

```typescript
export default function ActionButton ({ isBooked, id, isInCart }: ActionButtonProps): React.ReactNode {
  const [mutate, { loading, error }] = useMutation(
    isBooked ? CANCEL_TRIP : TOGGLE_CART,
    {
      variables: { launchId: id },
      refetchQueries: [
        {
          query: GET_LAUNCH_DETAILS,
          variables: { launchId: id },
        },
      ]
    }
  );

  if (loading) return <p>Loading...</p>;
  if (error) return <p>An error occurred</p>;

  return (
    <div>
      <Button
        onClick={() => mutate()}
        data-testid={'action-button'}
      >
        {isBooked
          ? 'Cancel This Trip'
          : isInCart
            ? 'Remove from Cart'
            : 'Add to Cart'}
      </Button>
    </div>
  );
}
```

trevorblades commented:
@peggyrayzis @hwillson yea, I've been moving towards function declarations as well. I like the way you can `export default function MyComponent()` and I think that `function handleEvent(event) { ...` looks nicer than the expression version. I still use arrow functions if I need to pass an anonymous function to a prop, for example:

```js
<button onclick={() => setCount(3)} />
```

stemmlerjs commented:
Function declarations it is! This might make the majority of code samples run off the page in the docs though.

stemmlerjs commented:
I just changed one component. Going to ask before I continue, **how do we feel about this**? Particularly the return type for _page_ components. 

When using `@reach/router`, in order to strictly type a page component using function declarations, we need:

- the `path?: string;` to be an attribute in the props for the component
- `props: CartProps` needs to be present in the function parameters even if we're not using it

> cart.tsx (with function declarations)

```typescript
import React, { Fragment } from 'react';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import { Header, Loading } from '../components';
import { CartItem, BookTrips } from '../containers';
import { RouteComponentProps } from '@reach/router';
import { GetCartItems } from './__generated__/GetCartItems';

export const GET_CART_ITEMS = gql`
  query GetCartItems {
    cartItems @client
  }
`;

interface CartProps {
  path?: string;
}

export default function Cart (
  props: CartProps
): React.ReactElement<RouteComponentProps<CartProps>> {
  
  const { data, loading, error } = useQuery<GetCartItems>(
    GET_CART_ITEMS
  );
  
  if (loading) return <Loading />;
  if (error) return <p>ERROR: {error.message}</p>;

  return (
    <Fragment>
      <Header>My Cart</Header>
      {!data || !!data && data.cartItems.length === 0 ? (
        <p data-testid="empty-message">No items in your cart</p>
      ) : (
        <Fragment>
          {!!data && data.cartItems.map((launchId: any) => (
            <CartItem key={launchId} launchId={launchId} />
          ))}
          <BookTrips cartItems={!!data ? data.cartItems : []} />
        </Fragment>
      )}
    </Fragment>
  );
}

```

Compare this with the previous version using function expressions.

> cart.tsx (with function expressions)

```typescript
import React, { Fragment } from 'react';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import { Header, Loading } from '../components';
import { CartItem, BookTrips } from '../containers';
import { RouteComponentProps } from '@reach/router';
import { GetCartItems } from './__generated__/GetCartItems';

export const GET_CART_ITEMS = gql`
  query GetCartItems {
    cartItems @client
  }
`;

interface CartProps extends RouteComponentProps {}

const Cart: React.FC<CartProps> = () => {
  const { data, loading, error } = useQuery<GetCartItems>(
    GET_CART_ITEMS
  );
  
  if (loading) return <Loading />;
  if (error) return <p>ERROR: {error.message}</p>;

  return (
    <Fragment>
      <Header>My Cart</Header>
      {!data || !!data && data.cartItems.length === 0 ? (
        <p data-testid="empty-message">No items in your cart</p>
      ) : (
        <Fragment>
          {!!data && data.cartItems.map((launchId: any) => (
            <CartItem key={launchId} launchId={launchId} />
          ))}
          <BookTrips cartItems={!!data ? data.cartItems : []} />
        </Fragment>
      )}
    </Fragment>
  );
}

export default Cart;
```

Thoughts @hwillson @trevorblades @peggyrayzis? If the React community is moving towards function declarations, I'd take the overhead of implementing function declarations with TypeScript anyways in order to move with them.

stemmlerjs commented:
Alright, folks. Because the TypeScript API for `@reach/router` is [pretty ugly right now](https://github.com/reach/router/issues/141), we're going to stick to the function expression version because that's what the TypeScript community seems to be most successful with using right now.

hwillson commented:
Sounds good @stemmlerjs - this PR looks great! There are a few changes that will need to be made for Apollo Client 3 (`@apollo/client`), but I can prep those changes in a PR once this is in place (and AC 3 is still in `beta`, and will be for another month at least). Thanks for working on this!

esemeniuc commented:
Looks lovely, a perhaps new year's gift?
